render/src/utils.rs

```diff
@@ -65,23 +65,79 @@ pub fn glue_tables_to_jpeg<'a>(
 }
 
 /// Removes potential invalid JPEG data from SWF DefineBitsJPEG tags.
-///
-/// SWF19 p.138:
-/// "Before version 8 of the SWF file format, SWF files could contain an erroneous header of 0xFF, 0xD9, 0xFF, 0xD8 before the JPEG SOI marker."
 /// These bytes need to be removed for the JPEG to decode properly.
-pub fn remove_invalid_jpeg_data(mut data: &[u8]) -> Cow<[u8]> {
-    // TODO: Might be better to return an Box<Iterator<Item=u8>> instead of a Cow here,
-    // where the spliced iter is a data[..n].chain(data[n+4..])?
-    if data.starts_with(&[0xFF, 0xD9, 0xFF, 0xD8]) {
-        data = &data[4..];
-    }
-    if let Some(pos) = data.windows(4).position(|w| w == [0xFF, 0xD9, 0xFF, 0xD8]) {
-        let mut out_data = Vec::with_capacity(data.len() - 4);
-        out_data.extend_from_slice(&data[..pos]);
-        out_data.extend_from_slice(&data[pos + 4..]);
-        out_data.into()
+pub fn remove_invalid_jpeg_data(data: &[u8]) -> Cow<[u8]> {
+    // SWF19 errata p.138:
+    // "Before version 8 of the SWF file format, SWF files could contain an erroneous header of 0xFF, 0xD9, 0xFF, 0xD8
+    // before the JPEG SOI marker."
+    // 0xFFD9FFD8 is a JPEG EOI+SOI marker pair. Contrary to the spec, this invalid marker sequence can actually appear
+    // at any time before the 0xFFC0 SOF marker, not only at the beginning of the data. I believe this is a relic from
+    // the SWF JPEGTables tag, which stores encoding tables separately from the DefineBits image data, encased in its
+    // own SOI+EOI pair. When these data are glued together, an interior EOI+SOI sequence is produced. The Flash JPEG
+    // decoder expects this pair and ignores it, despite standard JPEG decoders stopping at the EOI.
+    // When DefineBitsJPEG2 etc. were introduced, the Flash encoders/decoders weren't properly adjusted, resulting in
+    // this sequence persisting. Also, despite what the spec says, this doesn't appear to be version checked (e.g., a
+    // v9 SWF can contain one of these malformed JPEGs and display correctly).
+    // See https://github.com/ruffle-rs/ruffle/issues/8775 for various examples.
+
+    // JPEG markers
+    const SOF0: u8 = 0xC0; // Start of frame
+    const RST0: u8 = 0xD0; // Restart (we shouldn't see this before SOS, but just in case)
+    const RST7: u8 = 0xD7;
+    const SOI: u8 = 0xD8; // Start of image
+    const EOI: u8 = 0xD9; // End of image
+
+    let mut data: Cow<[u8]> = if data.starts_with(&[0xFF, EOI, 0xFF, SOI]) {
```

Comment:
nit: `strip_prefix` nicely fits here.

Reply:
I think once this tiny thing is addressed, this should be good to go?

Reply:
See: #9153